vcs/gitcmd/repo.go

```diff
@@ -161,105 +161,123 @@ func (r *Repository) ResolveTag(name string) (vcs.CommitID, error) {
+	if len(f.only) == 0 {
+		return true
 	}
-	behindAhead := strings.Split(strings.TrimSuffix(string(out), "\n"), "\t")
-	b, err := strconv.ParseUint(behindAhead[0], 10, 0)
-	if err != nil {
-		return 0, 0, err
+	_, ok := f.only[name]
+	return ok
+}
+
+// add adds a slice of strings to the filter.
+func (f *stringFilter) add(list []string) {
+	if f.only == nil {
+		f.only = make(map[string]struct{})
 	}
-	a, err := strconv.ParseUint(behindAhead[1], 10, 0)
-	if err != nil {
-		return 0, 0, err
+	for _, l := range list {
+		f.only[l] = struct{}{}
 	}
-	return uint(b), uint(a), nil
 }
 
 func (r *Repository) Branches(opt vcs.BranchesOptions) ([]*vcs.Branch, error) {
 	r.editLock.RLock()
 	defer r.editLock.RUnlock()
 
-	refs, err := r.showRef("--heads")
-	if err != nil {
-		return nil, err
-	}
-
-	branches := make([]*vcs.Branch, len(refs))
-	for i, ref := range refs {
-		branches[i] = &vcs.Branch{
-			Name: strings.TrimPrefix(ref[1], "refs/heads/"),
-			Head: vcs.CommitID(ref[0]),
+	var f stringFilter
+	if opt.MergedInto != "" {
+		b, err := r.branches("--merged", opt.MergedInto)
+		if err != nil {
+			return nil, err
 		}
+		f.add(b)
 	}
-
 	if opt.ContainsCommit != "" {
-		filteredBranchNames, err := r.branchesWithCommit(opt.ContainsCommit)
+		b, err := r.branches("--contains=" + opt.ContainsCommit)
 		if err != nil {
 			return nil, err
 		}
-		var filteredBranches []*vcs.Branch
-		for _, branch := range branches {
-			if _, in := filteredBranchNames[branch.Name]; in {
-				filteredBranches = append(filteredBranches, branch)
-			}
-		}
-		branches = filteredBranches
+		f.add(b)
+	}
+
+	refs, err := r.showRef("--heads")
+	if err != nil {
+		return nil, err
 	}
 
-	if opt.IncludeCommit {
-		for i, branch := range branches {
-			c, err := r.getCommit(branch.Head)
+	var branches []*vcs.Branch
+	for _, ref := range refs {
+		name := strings.TrimPrefix(ref[1], "refs/heads/")
+		id := vcs.CommitID(ref[0])
+		if !f.allows(name) {
+			continue
+		}
+
+		branch := &vcs.Branch{Name: name, Head: id}
+		if opt.IncludeCommit {
+			branch.Commit, err = r.getCommit(id)
 			if err != nil {
 				return nil, err
 			}
-			branches[i].Commit = c
 		}
-	}
-
-	// Fetch behind/ahead counts for each branch.
-	if opt.BehindAheadBranch != "" {
-		for i, branch := range branches {
-			behind, ahead, err := r.branchCounts(branch.Name, opt.BehindAheadBranch)
+		if opt.BehindAheadBranch != "" {
+			branch.Counts, err = r.branchesBehindAhead(name, opt.BehindAheadBranch)
 			if err != nil {
 				return nil, err
 			}
-			branches[i].Counts = &vcs.BehindAhead{Behind: uint32(behind), Ahead: uint32(ahead)}
 		}
+		branches = append(branches, branch)
 	}
-
 	return branches, nil
 }
 
-func (r *Repository) branchesWithCommit(commitID string) (map[string]struct{}, error) {
-	cmd := exec.Command("git", "branch", "--contains="+commitID)
+// branches runs the `git branch` command followed by the given arguments and
+// returns the output as a string.
```

Comment:
This method returns `[]string`, not `string` as the docs imply.

"and returns the output as a string slice"? Or just "returns the branches and error".